File: csharp/tournament/Tournament.cs
using System;
using System.IO;
using System.Collections.Generic;

public static class Tournament
{
    public static void Tally(Stream inStream, Stream outStream)
    {
        var reader = new StreamReader(inStream);
        var input = reader.ReadToEnd();

        var teams = new HashSet<string>();

        var played = new Dictionary<string, int>();
        var won = new Dictionary<string, int>();
        var drawn = new Dictionary<string, int>();
        var lost = new Dictionary<string, int>();
        var points = new Dictionary<string, int>();

        var lines = input.Split('\n');

        foreach (var line in lines) {
            // var (a, b, result) = line.Split(';');
            var tokens = line.Split(';');
            var a = tokens[0];
            var b = tokens[1];
            var result = tokens[2];

            teams.Add(a);
            teams.Add(b);

            played[a] += 1;
            played[b] += 1;

            switch (result) {
            case "win":
                won[a] += 1;
                lost[b] += 1;
                points[a] += 3;
                break;
            case "lose":
                lost[a] += 1;
                won[b] += 1;
                points[b] += 3;
                break;
            case "draw":
                drawn[a] += 1;
                drawn[b] += 1;
                points[a] += 1;
                points[b] += 1;
                break;
            }
        }

        var sorted = new string[teams.Count];
        teams.CopyTo(sorted);

        sorted.Sort((a, b) => {
            if (points[b] == points[a])
                return a.CompareTo(b);
            else
                return points[b].CompareTo(points[a]);
        });

        // var output = "Team                           | MP |  W |  D |  L |  P";

        var writer = StreamWriter(outStream);

        writer.WriteLine("Team                           | MP |  W |  D |  L |  P");

        foreach (var team in sorted) {
            writer.WriteLine("{0,-30} | {1,2} | {2,2} | {3,2} | {4,2} | {5,2}".Format(
                team, played[team], won[team], drawn[team], lost[team], points[team]
            ));
        }
    }
}

File: ruby/tournament/tournament.rb
require "set"

class Tournament
  def self.tally(input)
    teams = Set.new

    played = Hash.new(0)
    won = Hash.new(0)
    drawn = Hash.new(0)
    lost = Hash.new(0)
    points = Hash.new(0)

    input.each_line(chomp: true) do |row|
      a, b, result = row.split(";")

      next unless a

      teams << a
      teams << b

      played[a] += 1
      played[b] += 1

      case result
      when "win"
        won[a] += 1
        lost[b] += 1
        points[a] += 3
      when "draw"
        drawn[a] += 1
        drawn[b] += 1
        points[a] += 1
        points[b] += 1
      when "loss"
        lost[a] += 1
        won[b] += 1
        points[b] += 3
      end
    end

    sorted = teams.sort do |a, b|
      if (points[b] == points[a])
        a <=> b
      else
        points[b] <=> points[a]
      end
    end

    out = "Team                           | MP |  W |  D |  L |  P\n"

    sorted.each do |team|
      out += "%-30s | %2i | %2i | %2i | %2i | %2i\n" %
        [team, played[team], won[team], drawn[team], lost[team], points[team]]
    end

    return out
  end
end
